File: fushi/test/pages/popup_dict_css_url_rewrite_test.js
// 行为守卫：真跑 assets/popup/dict-media.js 的 constructDictCss，断言词典 CSS 里的
// 相对 url() 被重写到媒体通道（BUG-2147）。
//
// 词典样式表被内联成 <style> 注入弹窗文档，相对 URL 于是相对**弹窗文档**解析
// （Android 是 file:///android_asset/.../popup/，Windows/iOS 是 initialData 的
// opaque origin），与词典目录毫无关系 —— 不重写就永远 404。
const fs = require('fs');
const path = require('path');
const vm = require('vm');
const assert = require('assert');

const repoRoot = process.cwd();
const source = fs.readFileSync(
  path.join(repoRoot, 'assets', 'popup', 'dict-media.js'),
  'utf8',
);

const sandbox = { window: {}, console };
sandbox.globalThis = sandbox;
vm.createContext(sandbox);
vm.runInContext(source, sandbox);

const { constructDictCss } = sandbox;
assert.strictEqual(
  typeof constructDictCss,
  'function',
  'dict-media.js must expose constructDictCss',
);

const DICT = 'CamPron';
const enc = encodeURIComponent;

// 1. 裸名 url() -> image:// 媒体通道
{
  const out = constructDictCss('.sp{background:url(sound.png)}', DICT, '.s');
  assert.ok(
    out.includes(`image://?dictionary=${enc(DICT)}&path=${enc('sound.png')}`),
    `bare url() must be rewritten to the media channel, got: ${out}`,
  );
}

// 2. ?query 必须剥掉：导入侧按裸文件名入库（extract_css_url_names）
{
  const out = constructDictCss(
    '.sp{background:url("sprite.gif?version=5.0.287")}',
    DICT,
    '.s',
  );
  assert.ok(
    out.includes(`path=${enc('sprite.gif')}`),
    `?query must be stripped so the media key matches the import side, got: ${out}`,
  );
  assert.ok(
    !out.includes('version=5.0.287'),
    'the query string must not survive into the media path',
  );
}

// 3. 单引号 / 无引号 / 带空白三种写法都要认
{
  for (const css of [
    ".a{background:url('bg.png')}",
    '.a{background:url(bg.png)}',
    '.a{background:url(  bg.png  )}',
  ]) {
    const out = constructDictCss(css, DICT, '.s');
    assert.ok(
      out.includes(`path=${enc('bg.png')}`),
      `url() quoting variant not handled: ${css} -> ${out}`,
    );
  }
}

// 4. 带 scheme 的 / 协议相对的 URL 原样保留（不能被解析到词典目录里去）
{
  const css =
    '.b{background:url(https://dictionary.cambridge.org/x.png)}' +
    '.c{background:url(data:image/gif;base64,AAAA)}' +
    '.d{background:url(//cdn.example.com/y.png)}';
  const out = constructDictCss(css, DICT, '.s');
  assert.ok(out.includes('url(https://dictionary.cambridge.org/x.png)'), 'remote URL must be kept');
  assert.ok(out.includes('url(data:image/gif;base64,AAAA)'), 'data: URL must be kept');
  assert.ok(out.includes('url(//cdn.example.com/y.png)'), 'protocol-relative URL must be kept');
  assert.ok(!out.includes('image://'), `nothing here may reach the media channel: ${out}`);
}

// 4b. 前导 `/` 在 MDict 里表示「.mdd 根」，不是文件系统绝对路径 —— 剥掉它，与
//     <img src="/x.png"> 走的 normalizeDictMediaPath 同解。两条通道对同一个写法
//     必须解出同一个 media key，否则同一份资源在 CSS 里和在条目里指向两个 key。
{
  const out = constructDictCss('.a{background:url(/sub/logo.png)}', DICT, '.s');
  assert.ok(
    out.includes(`path=${enc('sub/logo.png')}`),
    `a leading slash is mdd-root-relative and must be stripped, got: ${out}`,
  );
}

// 5. 字体已在 Dart 侧内联成 data:（强制 CORS 子资源，媒体通道带不了 CORS 头），
//    到这里必须原样放过 —— 重写成 image:// 只会把它变成被 CORS 拒绝。
{
  const out = constructDictCss(
    "@font-face{font-family:'ico-c';src:url(data:font/woff;base64,d09GRg) format('woff')}",
    DICT,
    '.s',
  );
  assert.ok(
    out.includes('url(data:font/woff;base64,d09GRg)'),
    `an already-inlined font must pass through untouched, got: ${out}`,
  );
  assert.ok(!out.includes('image://'), 'a font must never be routed through image://');
}

// 6. 没有 url() 的 CSS 不受影响（早退路径）
{
  const out = constructDictCss('.a{color:red}', DICT, '.s');
  assert.ok(out.includes('color:red'), 'plain CSS must survive scoping');
}

console.log('all assertions passed');

File: fushi/test/pages/popup_dict_css_url_rewrite_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// BUG-2147：词典自带 CSS 里的相对 `url()` 必须被重写到媒体通道。
///
/// 样式表被**内联成 `<style>`** 注入弹窗文档，所以里面的相对 URL 是相对**弹窗文档**
/// 解析的（Android 是 `file:///android_asset/.../popup/`，Windows/iOS 是
/// `initialData` 的 opaque origin），与词典目录毫无关系 —— 不重写就永远 404，
/// 雪碧图/背景图/图标全部不显示。
///
/// 两道守卫：
/// 1) 行为 —— node 真跑 `assets/popup/dict-media.js` 的 `constructDictCss`
///    （见同名 .js）。node 不在 PATH 时跳过。
/// 2) 镜像 —— `dict-media.js` 实际有**三份**：
///      a. `fushi/assets/popup/dict-media.js`（app 内弹窗）
///      b. `tools/browser-extension/vendor/dict-media.js`（扩展源，是 a 的**超集**）
///      c. `fushi/assets/browser_extension/vendor/dict-media.js`（随包的扩展副本）
///    b↔c 由 `dart run fushi/tool/sync_browser_extension.dart` 保持字节一致，已有
///    守卫（`browser_extension_installer_test` 等三条）钉死。**a→b 没有任何工具覆盖**
///    （`sync-mirrors.mjs` 不做这一向），只能靠这里钉住：改了 a 忘了 b，扩展弹窗里的
///    词典背景图/雪碧图会全部 404，而 a 的测试照绿。
void main() {
  test('dictionary CSS url() is rewritten to the media channel (node)',
      () async {
    final String? nodeExe = _resolveNode();
    if (nodeExe == null) {
      markTestSkipped('node not found on PATH; skipping JS behavior execution');
      return;
    }

    final File jsTest = File('test/pages/popup_dict_css_url_rewrite_test.js');
    expect(jsTest.existsSync(), isTrue,
        reason: 'behavior harness ${jsTest.path} must exist');

    final ProcessResult result = await Process.run(
      nodeExe,
      <String>[jsTest.path],
      workingDirectory: Directory.current.path,
    );

    expect(
      result.exitCode,
      0,
      reason: 'dict CSS url() rewrite JS behavior test failed.\n'
          'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
    );
    expect(
      result.stdout.toString(),
      contains('all assertions passed'),
      reason: 'behavior harness must reach its success marker',
    );
  });

  test('both dict-media.js mirrors carry the url() rewrite', () {
    const List<String> mirrors = <String>[
      'assets/popup/dict-media.js',
      '../tools/browser-extension/vendor/dict-media.js',
    ];
    for (final String relative in mirrors) {
      final File file = File(relative);
      expect(file.existsSync(), isTrue, reason: '$relative must exist');
      final String source = file.readAsStringSync();
      expect(
        source.contains('function rewriteDictCssUrls('),
        isTrue,
        reason: '$relative 缺 url() 重写：该镜像里的词典背景图/雪碧图会全部 404',
      );
      expect(
        source.contains('css = rewriteDictCssUrls(css, dictName);'),
        isTrue,
        reason: '$relative 定义了 rewriteDictCssUrls 却没在 constructDictCssUncached '
            '里调用——定义不调用等于没写',
      );
    }
  });
}

String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) return name;
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}
